Modify Pipe to return an RRef.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48467 Update pipeline API to accept arbitrary sequence of Tensors and not just Tuple
* **#47829 Modify Pipe to return an RRef.**

As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

pritamdamania committed:
Modify Pipe to return an RRef.

As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/_pipeline/sync/pipe.py:
```diff
@@ -304,18 +305,18 @@ def _ensure_copy_streams(self) -> List[List[AbstractStream]]:
 
         Returns:
-            tensor or tensors: output mini-batch
+            ``RRef`` to the output of the mini-batch
```
nit: should it need to be `torch.distributed.rpc.RRef`?

rohan-varma commented on torch/testing/_internal/distributed/pipeline/utils.py:
```diff
@@ -4,8 +4,27 @@
+import tempfile
+
+@pytest.fixture
+def setup_rpc(scope="session"):
```
Looks like this function isn't used?

rohan-varma reviewed (approved):
LGTM, although there are lint errors.

pritamdamania committed:
Update on "Modify Pipe to return an RRef."


As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Modify Pipe to return an RRef."


As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Modify Pipe to return an RRef."


As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Modify Pipe to return an RRef."


As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Modify Pipe to return an RRef."


As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Modify Pipe to return an RRef."


As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Modify Pipe to return an RRef."


As per proposal in https://github.com/pytorch/pytorch/issues/44827,
the API needs to return an RRef to support inter-host pipelining.

For now, we just return a local RRef and only support pipeline on a single
host. But having this change in the API upfront ensures we don't make any BC
breaking changes later.

Differential Revision: [D24914022](https://our.internmc.facebook.com/intern/diff/D24914022/)

[ghstack-poisoned]
